File: 139.py
from sets import Set
class Solution(object):
    def dfs(self,s,l,check):
        if(s==""):
            return True;
        for ss in l:
            if(ss == s):
                return True
            if(len(ss) > len(s) or s[0:len(ss)] != ss): 
                continue;
            ts = s[len(ss)-len(s):]
            if(ts in check):
                continue
            found = self.dfs(ts,l,check)
            if(found == True):
                return True
        check.add(s)
        return False
    def wordBreak(self, s, wordDict):
        l = Set([])
        check = Set([])
        for ele in wordDict:
            l.add(ele)
        return self.dfs(s,l,check)

File: 139.cpp
class Solution {
public:
    bool dfs(const std::string& str, const std::vector<std::string>& dict){
        if(str == "") return true;
        if(visited.find(str) != visited.end()) return false;
        for(const std::string& s: dict){
            if(s.size() > str.size() || str.substr(0,s.size()) != s) continue;
            if(dfs(str.substr(s.size()), dict)) return true;
        }
        visited.insert(str);
        return false;
    }
    bool wordBreak(string s, vector<string>& wordDict) {
        return dfs(s,wordDict);
    }
private:
    std::unordered_set<std::string> visited;
};
